### ISSUE TITLE
[RFC]: Moving most torch.compile backends out of core

### ISSUE THREAD

**Issue by msaroufim** [pytorch/pytorch#109687]
## TL;DR

If you own a backend in https://github.com/pytorch/pytorch/tree/main/torch/_dynamo/backends and we don't hear back from you **we would like to delete it from core by 12/1/23** - if you strongly disagree please let us know ASAP and explain below why you feel the custom registration mechanism is not suitable

This will be BC breaking

I realize this date might be frustrating but I've given a heads up to all the POCs listed below since March of this year so without some due date I worry this won't be prioritized.

## Problem

When `torch.compile()` was first released it came with a `backend` parameter that made it really easy to try out different compilers. The problem was that 
1. Many backends were not tested in CI so it was unclear what their expected pass rate was on the HF, TIMM or torchbench suite
2. The backends may not be running on the right hardware relative to CI
3. Because those backends are not maintained by the pytorch core/compiler team we can't offer the same support model to fix bugs
4. The backends will fail by default pending some optional dependencies that we can't add as pytorch dependencies or pytorch CI so out of the box users still need to install extra things
5. There is a perception that backends in core are higher quality than backends out of core which may or may not be the case, we don't want to be in the business of blessing some backends at the expense of others

## Solution

So the solution we proposed was an out of core registration mechanism highlighted here https://pytorch.org/docs/main/torch.compiler_custom_backends.html#registering-custom-backends which meant that end users of pytorch could retain the convenience of `torch.compile(..., backend="...")` without having backend specific code living in core. In fact we believe that having the code out of core will create a superior backend since it will increase dev velocity and the testing infrastructure can and should mimic what your end users care about. For us a good proxy for what users care about has been a combination of huggingface, timm and torchbench models https://github.com/pytorch/pytorch/tree/main/benchmarks/dynamo

If registering a custom backend is too clunky then please provide feedback on this RFC and we can make it better, our goal should be to have little to no friction imposed on end users relative to the current in core backends. Remember that for an existing in core backend they will fail by default unless you install something.

If the goal was to promote a new backend, again there is no requirement that the code needs to live in `pytorch/pytorch` we in fact have a precedent for promoting research backends like Hidet that used an out of core registration mechanism 
 which we featured prominently on our blog and retweeted https://pytorch.org/blog/introducing-hidet/

So we want to really separate marketing vs technical reasons for code to be in core so the general principle for new dynamo/compile backends is if your code will be maintained by pytorch core/compiler engineers it should be in core and if it is not then the Linux foundation marketing team can help you out.

## Criteria to market a backend

Again we don't want to be in the business of blessing or endorsing backends, what matters most is what customers use and care about and not what I care about. So that end in order for us to market a backend there are some quality guidelines that will be helpful to follow so you can use this as a checklist before engaging with the Linux foundation

1. The backend has a clear list of maintainers
2. The backend has been maintained for at least 6 months
3. The pass rate is comparable to inductor, see https://hud.pytorch.org/benchmark/compilers
4. The perf speedups are comparable to inductor, see https://hud.pytorch.org/benchmark/compilers

You can leverage our benchmark scripts as a starting point to prove out 3,4 https://github.com/pytorch/pytorch/tree/main/benchmarks/dynamo and moving forward we'll use this criteria to for example include a backend here https://pytorch.org/docs/main/torch.compiler.html - Again this does not mean that your backend is pytorch blessed/approved/supported it's a just proxy for what we think is a good backend 

## Backend Tracker

We already have a precedent for not accepting new backends https://github.com/pytorch/pytorch/pull/104426 but the next step is making sure that all backends in core are maintained by the pytorch core/compiler team

These are the backends that we should move out of core

* [x] IPEX - this is registered out core  https://github.com/pytorch/pytorch/pull/104329 - owner @ashokei
* [x] NVfuser was already removed https://github.com/pytorch/pytorch/pull/105789 - owner @IvanYashchuk 
* [ ] ONNXRT - owner @abock 
* [ ] TensorRT - this is already a shim backend and can be removed in fact the out of core registration mechanism already exists https://github.com/pytorch/TensorRT/pull/2311 - owner @narendasan 
* [ ] #113382
* [ ] TVM - owner @shingjan

cc @ezyang @wconstab @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @malfet @gottbrath @gchanan

**Comment by thiagocrepaldi** [comment]
@abock FYI

**Comment by msaroufim** [comment]
@thiagocrepaldi @JackCaoG @shingjan just a reminder this is happening in less than a month

**Comment by wschin** [comment]
PyTorch is a complicated system. If `onnxrt` is moved out, it will become very unstable because new PyTorch commit can break it tomorrow.

**Comment by malfet** [comment]
@wschin we can keep running the tests, and if `onnxrt` is installed and register itself via extension point nothing will change for the end user

**Comment by wschin** [comment]
Well, it will also increase development cost -- as you split code into different repos -- because
1. it cannot enjoy the same build system as built-in backends.
2. more time to maintain different repos and their dependencies (this will be user's pain as well).

**Comment by malfet** [comment]
@wschin can you please elaborate on your points?  How it will be different from what it is right now?

**Comment by malfet** [comment]
> 3. torch.compile/inductor didn't even work with huggingface models before (while some blog posts already claimed its speedup to those models). There should be a grace period such as 6 months and a clear tolerance such as 10% difference pass rate is acceptable.

Assuming you are referring to `The pass rate is comparable to inductor` here, not sure how passing/failing hugging face models is relevant to this discussion here.

It would be good to define some more measurable criteria here, but  I think establishing that other pluggins are always lagging behind feels like a wrong incentive. 

> 4. This doesn't look good. E.g., how do you judge the speedup of XLA for TPU?

Performance gains are measured against default execution mode/eager vs further gains made via torch.compile. If there are no perf changes (nor say latency/memory utilization reduction) then what is the point of having a backend?

**Comment by wschin** [comment]
> > 3. torch.compile/inductor didn't even work with huggingface models before (while some blog posts already claimed its speedup to those models). There should be a grace period such as 6 months and a clear tolerance such as 10% difference pass rate is acceptable.
> 
> Assuming you are referring to `The pass rate is comparable to inductor` here, not sure how passing/failing hugging face models is relevant to this discussion here.
> 
> It would be good to define some more measurable criteria here, but I think establishing that other pluggins are always lagging behind feels like a wrong incentive.
> 
> > 4. This doesn't look good. E.g., how do you judge the speedup of XLA for TPU?
> 
> Performance gains are measured against default execution mode/eager vs further gains made via torch.compile. If there are no perf changes (nor say latency/memory utilization reduction) then what is the point of having a backend? (Unless it enables a completely new HW platform)

Not only new HW but other OS such as Windows needs to be considered when setting up experiments for speed comparison. In some cases, you don't even have CUDA and eager mode runs on CPU...

**Comment by wschin** [comment]
> @wschin can you please elaborate on your points? How it will be different from what it is right now?

I am having my lunch, so just a quick one.

Think about a bug needs joint behavior change in PyTorch Core and onnxrt ---

1. If I change onnxrt first, PyTorch CI breaks.
2. If I change PyTorch Core first, PyTorch CI breaks.
3. If I change PyTorch Core and ping onnxrt to a dev commit, CI passes but it leads to security concerns. It also increases the dev cost since, if I ping PyTorch to wrong commit, it will take half hour to rebuild.

The dependency is a nightmare as well. When dev or user hits a bug, now they have to try all combination of PyTorch, onnxrt, onnxruntime to narrow the scope of the bug. Some python installation process triggers strange cache behavior and adding one extra python package will complicate the problem a lot.

**Comment by thiagocrepaldi** [comment]
I appreciate the ongoing discussions around the proposal to remove third-party backends from the PyTorch repository, and I would like to contribute some technical considerations that I believe merit consideration.
 
From a technical standpoint, my concerns primarily revolve around the discoverability of "non-core" backends, which plays a vital role in user familiarity and adoption. Presently, users can explore existing backends through the [torch.compiler doc page](https://pytorch.org/docs/stable/torch.compiler.html), and programmatically access this information via `torch.compiler.list_backends()` API. However, it appears that the proposed RFC might result in the removal of references to these backends if they are relocated outside the PyTorch repository.
 
If this backend removal proposal advances, new users may only become aware of the new/most popular or well-maintained `torch.compile` backends through individual efforts or in collaboration with Linux Foundation/PyTorch marketing initiatives. Although this is a great resource for when all backends are up to speed with all the marvels of Dynamo, this could pose a potential barrier to adoption, as the existing method of discovering backends through `torch.compiler.list_backends()` would only list `inductor`.
 
Recognizing that `torch.compiler.list_backends` may currently be considered a less critical API, possibly due to popular backends already residing in the PyTorch repo or because backend developers are in early stages of development, I propose considering an enhancement to this API or another alternative approach.
 
My suggestion is to maintain in the PyTorch core a curated list of select backends, complete with references to URLs and short descriptions, accessible through `torch.compiler.list_backends`. This list would include both in-source and out-of-source backends, offering new users an official discovery point before installing additional dependencies. The backend list would be based on a to-be-defined clear technical criteria for inclusion and, factors such as pass rate, latency, and release frequency could be considered.
 
An associated change would involve extending `torch.compile(backend=...)` error handling to provide informative messages when a backend is not installed but is part of the select backend list. For example, the message could read: "This backend is supported by torch.compile but it is not installed. Refer to https://... for instructions on how to get started". On the other hand, non-select backends could raise an error directing users to `torch.compiler.list_backends` for information on available backends.
 
Additionally, to boost discoverability, keeping the most popular backends in the official PyTorch documentation (e.g., accessible through https://pytorch.org/docs, https://pytorch.org/backends, and/or https://pytorch.org/docs/stable/torch.compiler.html) could further aid users in exploring available options.
 
I view this proposal as a complementary measure that aims to bridge the gap between core and external backends, should the decision be made to proceed with the proposed separation. I believe these enhancements would not only maintain a smooth user experience but also foster a more inclusive and discoverable ecosystem for PyTorch users.

**Comment by msaroufim** [comment]
@thiagocrepaldi your compromise sounds like a good one and it's one @malfet and I have discussed as well, I'm supportive

@JackCaoG @shauheen sounds good to you?

**Comment by malfet** [comment]
Thank you for a thoughtful  write up @thiagocrepaldi 

I think having some mechanism that would help discoverability of available, but yet not installed plugins would be very helpful, but at the same time I don't think  PyTorch project should be responsible for maintaining the list of blessed backends in perpetuity. 
Similarly,  Python Foundation has no say in what is and what is not getting published in pypi, PyTorch in the long run should be better of by avoiding passing any judgement on backend developers. (I.e. there should be a very simple form to register a backend and it later should be mentioned in one of the lists)

In the short term, I suggest we convert list of backends mentioned in [torch.compielr](https://pytorch.org/docs/main/torch.compiler.html) into a yaml file, which will be used to generate a docs as well as some sort of help message if `torch.compile(backend="xyz")` is called but backend `xyz` has not been registered via extension point.

> My suggestion is to maintain in the PyTorch core a curated list of select backends, complete with references to URLs and short descriptions, accessible through torch.compiler.list_backends.

It must be a different API as currently `list_backends` returns a list of strings, which is not very informative (and IMO quite confusing)
```
>>> import torch
>>> print(torch.compiler.list_backends.__doc__)

    Return valid strings that can be passed to `torch.compile(..., backend="name")`.

    Args:
        exclude_tags(optional): A tuple of strings representing tags to exclude.
    
>>> torch.compiler.list_backends()
['cudagraphs', 'inductor', 'onnxrt', 'openxla', 'openxla_eval', 'tvm']
```


> An associated change would involve extending torch.compile(backend=...) error handling to provide informative messages when a backend is not installed but is part of the select backend list. For example, the message could read: "This backend is supported by torch.compile but it is not installed. Refer to https://... for instructions on how to get started". On the other hand, non-select backends could raise an error directing users to torch.compiler.list_backends for information on available backends.

This sounds great, though imo message should be more neutral, as right now I have no idea what statement "Backed XYZ is supported by torch.compile" actually implies.

**Comment by shauheen** [comment]
Thanks @malfet @msaroufim and @thiagocrepaldi for the great discussion and suggestions. I want to echo the concern around discoverability. This RFC as proposed will certainly have a negative impact on the backends that are not developed by the primary team at Meta and is putting `inductor` and related backends at an advantage point through ease of discoverability and access.

I really think this RFC should not be implemented as is proposed. I don't think the code in question is significant enough such that their removal would have any meaningful impact on code health maintenance. 

If the concern is the quality of the backends that are listed, I propose that we define a simple process of how long before the maintaining team would resolve an issue filed against their backend. If the backend maintainers have not resolved any issues between two releases, then the code for those backend should be removed from the API.

> PyTorch in the long run should be better of by avoiding passing any judgement on backend developers

This way no one even has to be passing any judgement on backend developers, and it simply is a question of whether the maintainer is actively responsive or has moved on from the backend. The choice of using or not using a backend will always remain with the end users.

**Comment by wschin** [comment]
> Thanks @malfet @msaroufim and @thiagocrepaldi for the great discussion and suggestions. I want to echo the concern around discoverability. This RFC as proposed will certainly have a negative impact on the backends that are not developed by the primary team at Meta and is putting `inductor` and related backends at an advantage point through ease of discoverability and access.
> 
> I really think this RFC should not be implemented as is proposed. I don't think the code in question is significant enough such that their removal would have any meaningful impact on code health maintenance.
> 
> If the concern is the quality of the backends that are listed, I propose that we define a simple process of how long before the maintaining team would resolve an issue filed against their backend. If the backend maintainers have not resolved any issues between two releases, then the code for those backend should be removed from the API.
> 
> > PyTorch in the long run should be better of by avoiding passing any judgement on backend developers
> 
> This way no one even has to be passing any judgement on backend developers, and it simply is a question of whether the maintainer is actively responsive or has moved on from the backend. The choice of using or not using a backend will always remain with the end users.

+1. In addition to discoverability and access, as explained above, this will also increase development cost to non-Meta backends. As different backends now competing against each other in some aspects, it's important to maintain the fairness for all developers and companies since PyTorch is governed under Linux Foundation.

**Comment by laikhtewari** [comment]
+1 to the great points made on impacts to discoverability. I think this RFC also hurts the PyTorch UX since users must manually intervene to register non-Inductor backends (which isn't solved by the proposal to change `list_backends` behavior)

**Comment by msaroufim** [comment]
>  I want to echo the concern around discoverability. This RFC as proposed will certainly have a negative impact on the backends that are not developed by the primary team at Meta and is putting inductor and related backends at an advantage point through ease of discoverability and access.

It seems like the core of the argument is that code in-core has a disadvantage when it comes to visibility relative to code out of core but this argument is not correct. There are many libraries on top of PyTorch that have built substantial followings including Huggingface, Deepspeed, Lightning, PyTorch Geometric and many **many** more. All those libraries have in common that they build on top of PyTorch and are not merged inside PyTorch. But most importantly they are beloved by their users and what ultimately determines whether a backend is successful is not PyTorch developers but PyTorch users.

> +1 to the great points made on impacts to discoverability. I think this RFC also hurts the PyTorch UX since users must manually intervene to register non-Inductor backends (which isn't solved by the proposal to change list_backends behavior)

Users still need to pip install some dependency without this RFC, even then:

I'm not convinced that `list_backends()` is the way new users will discover backends. PyTorch users will "vote" by opening issues, contributing fixes, asking to have outside code merged, writing tutorials, and publishing case studies. You could make a similar argument that perhaps PyTorch should merge some new optimizer for example and generally our policy is if we feel like a substantial chunk of our users are using some feature then it might make sense to merge some code into core.

> +1. In addition to discoverability and access, as explained above, this will also increase development cost to non-Meta backends. 

I would like to hear a **technical** argument for why this is the case, I linked to how to register out of core backends and this was performed by many organizations including NVIDIA, Intel, and Hidet none of which mentioned an increased development cost and their PRs are a simple blueprint for how to do so.

**Comment by msaroufim** [comment]
We can delay the decision but I'd still like to see someone address my central points https://github.com/pytorch/pytorch/issues/109687#issuecomment-1823779802

**Comment by wschin** [comment]
@msaroufim , a previous comment https://github.com/pytorch/pytorch/issues/109687#issuecomment-1804640507 has a technical concern about harder dynamo-and-dynamo-backend bug fix and harder version control debugging. I don't recently work on dynamo, so I probably forget a lot of details already. If it's possible, maybe you can get some feedback from Meta's internal engineers about what would be the extra development cost if inductor is moved out. I feel the situation would be similar.
